### ISSUE TITLE
Allow users to add free-form text to qubes (for descriptions, notes, comments, remarks, reminders, etc.)

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#899]
**Reported by Oleg Artemiev  grey olli on 12 Sep 2014 08:14 UTC**
It could be usefull to have a text description for a VM in its
properties to document changes/difference.
https://groups.google.com/d/topic/qubes-users/uYCcOqIeF5k/discussion

Migrated-From: https://wiki.qubes-os.org/ticket/899

**Comment by Jeeppler** [comment]
I would like to work on this issue (see https://groups.google.com/forum/#!topic/qubes-devel/t32l-0BjlLs).

**Comment by marmarek** [comment]
@bnvk is working on new Qubes Manager, please coordinate.

**Comment by Jeeppler** [comment]
Is this still relevant, after [decomposing the Qubes Manager](https://github.com/QubesOS/qubes-issues/issues/2132)?

**Comment by marmarek** [comment]
On Tue, Jul 26, 2016 at 02:10:35PM -0700, Jeppler wrote:
> Is this still relevant, after [decomposing the Qubes Manager](https://github.com/QubesOS/qubes-issues/issues/2132)? How would it look like? Would it be something like a `qvm-description` tool? I would love to have a VM description.

I think this may be simply a VM property, accessible using qvm-prefs.
How to present it in GUI is unrelated to this ticket.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by Jeeppler** [comment]
Perfect. @marmarek are you considering to implement it as qvm-prefs property in Q R4?

**Comment by andrewdavidwong** [comment]
It's worth thinking about how this is related to #865, i.e., defining and disambigutating these possible properties of VMs:

* Description
* Tag
* Attribute

**Comment by Jeeppler** [comment]
I would like to see "Tags" the plural form to do something like: <project-name>-dev -> tags: work, project

It would be nice to have a "last started" property. This would be helpful to see if the machines are used regularly or have been used a long time ago.

**Comment by Jeeppler** [comment]
What would be the "Attribute" property? It seems a little bit generic.

**Comment by andrewdavidwong** [comment]
Yes, "last started" would be useful, especially when trying to decide whether a VM needs to be included in a backup. (If the "last started" date is the same or earlier than the "last backup" date, then it probably doesn't need to be backed up again.)

**Comment by marmarek** [comment]
> I would like to see "Tags" the plural form to do something like: -dev -> tags: work, project

Yes, that's how it is in Qubes 4.0.

> What would be the "Attribute" property? It seems a little bit generic.

That's generic name for tag with some value (so, not only "present"/"absent") ;)

**Comment by marmarek** [comment]
Some thoughts about "VM description" feature - it would be useful to allow free-form text, possibly with some non-ASCII characters, multiple lines, etc. So I take back the idea of putting it in qvm-prefs (and consequently in qubes.xml), as it may lead to some troubles (qvm-prefs don't support multi-line values for example). IMO much better store it as a simple `description.txt` file in VM directory (`/var/lib/qubes/appvms/name-of-vm/description.txt`). And in preparation for GUI domain - have some qrexec services to access it (`qubes.DescriptionGet`, `qubes.DescriptionSet`). For now, those services would be called from dom0 (this is where qubes-manager is running) to dom0.

**Comment by andrewdavidwong** [comment]
Would that be a potential attack vector in conjunction with `qvm-export`?

**Comment by marmarek** [comment]
What do you have in mind? Trying to exploit archive format from inside of included file? That archive tool should be resistant against such attacks because it needs to include also `private.img`, which is fully controlled by the VM.

**Comment by andrewdavidwong** [comment]
Well, we disallow rendering non-ASCII text in dom0 window titles by default, since window title content is potentially adversary-controlled and therefore presents a potential attack vector against dom0. IIUC, `qvm-export` is meant to allow importing potentially untrusted VMs. But if we allow non-ASCII characters in `description.txt`, which also gets imported and rendered in dom0, then we've reintroduced adversary-controlled non-ASCII text rendered in dom0. Is that a problem?

**Comment by Jeeppler** [comment]
@marmarek `it would be useful to allow free-form text, possibly with some non-ASCII characters, multiple lines, etc.` yes, I absolutely agree with you.

@andrewdavidwong I assume the reason for only allowing rendering a specific subset of ASCII characters has to do with technical limitations of the window manager.

I would like to have Unicode (UTF-8) support. Unicode (UTF-8) allows to write text and words in almost all languages and is not limited to English.

**Comment by andrewdavidwong** [comment]
> @andrewdavidwong I assume the reason for only allowing rendering a specific subset of ASCII characters has to do with technical limitations of the window manager.
> 
> I would like to have Unicode (UTF-8) support. Unicode (UTF-8) allows to write text and words in almost all languages and is not limited to English.

No, it's not a technical limitation. Rather, UTF-8 in window titles is disabled by default as a security measure for the reasons I mentioned above. If you like, you can enable UTF-8 in window titles by setting `allow_utf8_titles = true`, either globally or on a per-VM basis, in `/etc/qubes/guid.conf` in dom0.

**Comment by marmarek** [comment]
> But if we allow non-ASCII characters in description.txt, which also gets imported and rendered in dom0, then we've reintroduced adversary-controlled non-ASCII text rendered in dom0. Is that a problem?

Ah, indeed this may be a problem. I see a few options:
 - don't import description.txt as part of the VM importing
 - strip it of dangerous characters (which also will include `<`, `&` etc, to avoid triggering some HTML or other parsers)

**Comment by rootkovska** [comment]
1. I like keeping description as a separate file, not to pollute the XML file.

2. I think coming up with a general method to sanitize `description` would be tricky. We don't know what widget/application will finally want to display it. This is potentially much harder to reasonable sanitation of titlebars.

3. So, instead, I would go for ignoring it while... well that's the challenge: we can't just say: ignore it _only_ for imported VMs (restoring VMs previously qvm-exported by someone else), while retain for those imported from a backup, easily. Can we?

4. We should aim at treating any `qvm-backup-restore` operation as potentially untrusted. The ultimate goal for qvm-backup-restore should be that, even if the backup was made on a compromised Qubes system, than the new system (where we restoring to) should not get compromised -- i.e. neither Dom0, nor the hypervisor, and firmware, etc, should not be affected (this of course assumes the user opting out from restoring the Dom0 home dir, which often can be used with minimal PITA).

BTW just to double check, @andrewdavidwong: the scenario you really mean is qvm-backup-restore importing a previously qvm-exported (by someone else) VM, correct?

**Comment by marmarek** [comment]
On Sat, Apr 01, 2017 at 04:32:15AM -0700, Joanna Rutkowska wrote:
> 3. So, instead, I would go for ignoring it while... well that's the challenge: we can't just say: ignore it _only_ for imported VMs (restoring VMs previously qvm-exported by someone else), while retain for those imported from a backup, easily. Can we?

Yes, we can simply ignore this file during VM import - this will be a
separate tool, so it can have separate logic for handling this.

-- 
Best Regards,
Marek Marczykowski-Górecki
Invisible Things Lab
A: Because it messes up the order in which people normally read text.
Q: Why is top-posting such a bad thing?

**Comment by andrewdavidwong** [comment]
> BTW just to double check, @andrewdavidwong: the scenario you really mean is qvm-backup-restore importing a previously qvm-exported (by someone else) VM, correct?

*If* `qvm-backup-restore` is the tool used to import VMs that have been created (and potentially shared) with the `qvm-export-vm` tool (#1747), then yes. However, #1747 doesn't specify whether it would be `qvm-backup-restore` or a new tool, e.g., `qvm-import-vm`.

> Yes, we can simply ignore this file during VM import - this will be a
separate tool, so it can have separate logic for handling this.

Ok, sounds like it would indeed be a new tool (e.g., `qvm-import-vm`).

**Comment by rootkovska** [comment]
> Yes, we can simply ignore this file during VM import - this will be a
separate tool, so it can have separate logic for handling this.

But the doesn't solve the problem that we still want to have qvm-backup-restore to be resistant to a compromise if the previous system (on which qvm-bakup/qvm-export were run) was compromised.

Perhaps a solution to this would be to add a switch to qvm-backup-restore: e.g. `--distrust-backup", which would imply: 1) don't restore Dom0 homedir, 2) ignore VM description files, 3) perhaps some other things, TBD in another ticket.

And, yes, having a separate qvm-import tool, which always ignores description.txt files.

**Comment by iamahuman** [comment]
Simplest way would be to use a feature property.

**Comment by marmarek** [comment]
See https://github.com/QubesOS/qubes-issues/issues/899#issuecomment-290857248 why it isn't a good idea.

**Comment by andrewdavidwong** [comment]
(Increased priority and removed "help wanted" since they were last set many years ago, and there seems to be noticeable demand for this feature. This is just to prompt reevaluation, not any kind of authoritative decision. @marmarek, feel free to put them back.)

**Comment by HerzogM** [comment]
I really hope this will be addressed. I keep finding myself wanting to have a textbox in a qube’s settings to make remarks, comments or a description for that qube. Things I want to remember for that qube, a description for how this qube is intended to be used, warnings, hints etc.
The name field for the qube is not sufficient for this purpose and it would result in too long qube names if I made the name too verbose. So maybe the devs can consider giving the settings a textarea for this purpose?

**Comment by DemiMarie** [comment]
Assigning to @marmarta because the vast majority of the work here is UX related.  The backend part (saving stuff in `qubes.xml`) is trivial in comparison.

**Comment by alimirjamali** [comment]
Development notes:

- [x] Decide on the actual note file(s) name(s), format and their locations
- [ ]  Sanitization of (UTF) text on loading and saving (currently ASCII)
- [ ] Finalize decision on API calls (for example do we need sha256sum of notes via API?)
- [ ] Backup and Restore for notes
- [ ] Unit tests
- [ ] Update API documentation
- [ ] Notes cloning (while cloning a qube)
- [x] UX

**Comment by alimirjamali** [comment]
> Some thoughts about "VM description" feature - it would be useful to allow free-form text, possibly with some non-ASCII characters, multiple lines, etc. So I take back the idea of putting it in qvm-prefs (and consequently in qubes.xml), as it may lead to some troubles (qvm-prefs don't support multi-line values for example). IMO much better store it as a simple `description.txt` file in VM directory (`/var/lib/qubes/appvms/name-of-vm/description.txt`). And in preparation for GUI domain - have some qrexec services to access it (`qubes.DescriptionGet`, `qubes.DescriptionSet`). For now, those services would be called from dom0 (this is where qubes-manager is running) to dom0.

As per discussions on Matrix, the file name decided to be `notes.txt` rather than `description.txt`. And the location is as advised above. The API calls are decided to be `admin.vm.notes.Get` and `admin.vm.notes.Set`.

> Well, we disallow rendering non-ASCII text in dom0 window titles by default, since window title content is potentially adversary-controlled and therefore presents a potential attack vector against dom0. IIUC, `qvm-export` is meant to allow importing potentially untrusted VMs. But if we allow non-ASCII characters in `description.txt`, which also gets imported and rendered in dom0, then we've reintroduced adversary-controlled non-ASCII text rendered in dom0. Is that a problem?

The current draft only allows printable ASCII characters in addition to space, tab and newline. As per discussions on Matrix, a Python binding of the recent UTF-8 sanitisation mechanism will be added and will be used for notes. It is also agreed that no Markdown or HTML tag support should be added.

> 2\. We don't know what widget/application will finally want to display it.

[QPlainTextEdit](https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QPlainTextEdit.html) is used to render and display it. Unlike its sibling (QTextEdit), it does not support HTML or Markdown.

> 3\. So, instead, I would go for ignoring it while... well that's the challenge: we can't just say: ignore it _only_ for imported VMs (restoring VMs previously qvm-exported by someone else), while retain for those imported from a backup, easily. Can we?

We can make the import optional?

> 4\. We should aim at treating any `qvm-backup-restore` operation as potentially untrusted. The ultimate goal for qvm-backup-restore should be that, even if the backup was made on a compromised Qubes system, than the new system (where we restoring to) should not get compromised -- i.e. neither Dom0, nor the hypervisor, and firmware, etc, should not be affected (this of course assumes the user opting out from restoring the Dom0 home dir, which often can be used with minimal PITA).

The recent (UTF-8) sanitisation is exactly optimized for such scenarios and hopefully could be reused for this as well.

> Perhaps a solution to this would be to add a switch to qvm-backup-restore: e.g. `--distrust-backup", which would imply: 1) don't restore Dom0 homedir, 2) ignore VM description files, 3) perhaps some other things, TBD in another ticket.

Good

**Comment by alimirjamali** [comment]
We need a CLI tool to read & modify notes as well (outside of dom0). We could name it `qvm-notes`

- `qvm-notes VM` prints qube notes
- `qvm-notes [-e|--edit] VM` opens qube notes in `$EDITOR`
- `qvm-notes [-i|--import] VM [notes|-]` allows user to change qube notes from command line or pipe.
- `qvm-notes [-d|--delete] VM` truncates qube notes (sets it to "")
- `qvm-notes --append VM [notes|-]` allos user to append to qube notes from command line or pipe.

All of the above will be `admin.vm.notes.Get` and `admin.vm.notes.Set` API calls. No individual API calls is necessary.

**Comment by ben-grande** [comment]
> We need a CLI tool to read & modify notes as well (outside of dom0). We could name it qvm-notes

Example for editor https://github.com/QubesOS/qubes-core-qrexec/blob/main/qrexec/tools/qubes_policy_editor.py

**Comment by marmarek** [comment]
> IMO, the notes should not be readable and modifiable outside of dom0. Individual qubes should certainly not be able to read and modify their own notes.

This is of course subject to Admin API policy. Yes, by default qube cannot access any notes (own or others). But it can be allowed. This is relevant for example for sys-gui.

**Comment by tasket** [comment]
If this info is saved as individual files, one per VM, that would be ideal for backup & restore functions.

**Comment by alimirjamali** [comment]
> If this info is saved as individual files, one per VM, that would be ideal for backup & restore functions.

Notes will be indeed saved as individual files (notes.txt), one per VM. You could see the code for ordinary traditional backup [here](https://github.com/QubesOS/qubes-core-admin/pull/656/files#diff-76b331681aeac2a25d40eea44d69905dc19d4f31837ae696948089e002831076R436-R438). I hope it should be easy to adapt it for Wyng as well.

**Comment by alimirjamali** [comment]
Ok. The core-admin API patch is merged (this is why the @qubesos-bot closed the issue). The issue has to be reopened.

Time to finish the core-admin-client PR
